src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Tree/Dependencies/Subscriptions/RuleHandlers/DependenciesRuleHandlerBase.cs

```diff
@@ -51,62 +51,47 @@ internal abstract class DependenciesRuleHandlerBase
         public abstract ImageMoniker ImplicitIcon { get; }
 
         public void Handle(
-            IImmutableDictionary<string, IProjectChangeDescription> changesByRuleName,
+            IProjectChangeDescription evaluation,
+            IProjectChangeDescription? projectBuild,
             TargetFramework targetFramework,
             DependenciesChangesBuilder changesBuilder)
         {
-            // We receive evaluated and resolved project data separately, each as its own rule.
-
-            // We always have evaluated data.
-            IProjectChangeDescription evaluatedChanges = changesByRuleName[EvaluatedRuleName];
-
             HandleChangesForRule(
                 resolved: false,
-                projectChange: evaluatedChanges,
+                projectChange: evaluation,
                 isEvaluatedItemSpec: null);
 
             // We only have resolved data if the update came via the JointRule data source.
-            if (changesByRuleName.TryGetValue(ResolvedRuleName, out IProjectChangeDescription resolvedChanges))
+            if (projectBuild != null)
             {
-                Func<string, bool>? isEvaluatedItemSpec = ResolvedItemRequiresEvaluatedItem
-                    ? evaluatedChanges.After.Items.ContainsKey
-                    : (Func<string, bool>?)null;
+                Func<string, bool>? isEvaluatedItemSpec = ResolvedItemRequiresEvaluatedItem ? evaluation.After.Items.ContainsKey : null;
 
                 HandleChangesForRule(
                     resolved: true,
-                    projectChange: resolvedChanges,
+                    projectChange: projectBuild,
                     isEvaluatedItemSpec);
             }
 
             return;
 
             void HandleChangesForRule(bool resolved, IProjectChangeDescription projectChange, Func<string, bool>? isEvaluatedItemSpec)
             {
-                if (projectChange.Difference.RemovedItems.Count != 0)
+                foreach (string removedItem in projectChange.Difference.RemovedItems)
```

Comment:
The empty checks were added to avoid allocating enumerators.

Reply:
Oh I see your comment:

> On empty collections this returns a cached enumerator.

Reply:
Yes, this just returns Enumerable.Empty<T>(), which returns T[], which itself returns a cached enumerator for empty.